modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -3969,4 +4012,52 @@ public CacheRebalanceOrder(int order, CacheRebalanceMode mode) {
             return result;
         }
     }
+
+    /**
+     * Represents reassign context and stores node ids that cannot be used for historical rebalance for the given exchange.
+     */
+    private static class RebalanceReassignContext {
```

Comment:
I think RebalanceReassignContext is not needed.
Just keep a set of failed historical supplier ids until the node has left.

Reply:
I would prefer the idea to keep this set of nodes in the corresponding exchange future. This approach gives a try to the excluded node to become a historical supplier in the future.